### DISCUSSION TITLE
core(private-modules): ssh-agent forward seems to break when multiple keys are loaded

### DISCUSSION THREAD

**Discussion by grouville** [dagger/dagger#12236] [maintainers]
Following a discussion with @aweris https://discord.com/channels/707636530424053791/1279115372116705422/1279197286903320639

I don't think this is a socket forward issue, but more a config on our end, either on the tag / either on the branch discovery.

Digging into that

**Comment by grouville** [comment]
@aweris Is it still the case ? Our test suite has two keys loaded in the ssh agent and seems to work

**Comment by aweris** [comment]
I just revoked the second ssh key due to this issue, not sure if still exist or not

**Comment by ielsharef** [comment]
Hi there, I believe the issue I'm running into is related to this. I'm calling a private dagger module that takes as input a url to another private GitHub repo. The input type is *dagger.Directory since I'm looking to clone that second private repo for use by my private module.

Dagger version
dagger v0.16.1 (registry.dagger.io/engine:v0.16.1) linux/arm64
Go SDK

Steps to reproduce
1. Create 2 private repos; one will contain the dagger module and the other will be the repo that's passed as an arg
2. Add GitHub deploy keys to each private repo
3. Ensure ssh keys are set locally or in calling pipeline
4. Make dagger call

Example call
dagger -m git@github.com:private-daggerverse.git/modulex@0.30.2 call \\
    functiony \\
        --repo-to-clone=git@github.com:private-repo-to-clone.git

The result I see is that the key required for the second repo (one being passed in) doesn't get used correctly so the repo doesn't get cloned. I've tried passing the repo url as a string along with a socket arg to then clone using dag.git() within the module, but no luck.

[Clearing all loaded keys](https://docs.dagger.io/api/remote-modules/#multiple-ssh-keys-may-cause-ssh-forwarding-to-fail) doesn't work either because two keys are required to make this call. My workaround right now is to pass the url as a string arg along with an ssh key as a secret, and just have separate logic in Go to handle cloning the repo.

Do you have any recommendations I can try for calling a module / cloning natively with multiple ssh keys or will this just be a known limitation? Thanks!

**Comment by ielsharef** [comment]
@aweris @grouville bumping this, am I maybe posting in the wrong spot for this question? 🙂

**Comment by grouville** [comment]
> Hi there, I believe the issue I'm running into is related to this. I'm calling a private dagger module that takes as input a url to another private GitHub repo. The input type is *dagger.Directory since I'm looking to clone that second private repo for use by my private module.
> 
> Dagger version dagger v0.16.1 (registry.dagger.io/engine:v0.16.1) linux/arm64 Go SDK
> 
> Steps to reproduce
> 
> 1. Create 2 private repos; one will contain the dagger module and the other will be the repo that's passed as an arg
> 2. Add GitHub deploy keys to each private repo
> 3. Ensure ssh keys are set locally or in calling pipeline
> 4. Make dagger call
> 
> Example call dagger -m [git@github.com](mailto:git@github.com):private-daggerverse.git/modulex@0.30.2 call \ functiony \ --repo-to-clone=[git@github.com](mailto:git@github.com):private-repo-to-clone.git
> 
> The result I see is that the key required for the second repo (one being passed in) doesn't get used correctly so the repo doesn't get cloned. I've tried passing the repo url as a string along with a socket arg to then clone using dag.git() within the module, but no luck.
> 
> [Clearing all loaded keys](https://docs.dagger.io/api/remote-modules/#multiple-ssh-keys-may-cause-ssh-forwarding-to-fail) doesn't work either because two keys are required to make this call. My workaround right now is to pass the url as a string arg along with an ssh key as a secret, and just have separate logic in Go to handle cloning the repo.
> 
> Do you have any recommendations I can try for calling a module / cloning natively with multiple ssh keys or will this just be a known limitation? Thanks!

Hey @ielsharef – really appreciate the detailed explanation 🙏 Super helpful.

From what I understand, the issue comes up when:
- You’re calling a **private module** from one GitHub repo (using SSH key A),
- And that module is trying to clone another **private repo** that needs a **different SSH key B**.

You mentioned:
> "Clearing all loaded keys doesn't work either because two keys are required to make this call."

So it sounds like both keys are loaded into your `ssh-agent`, which makes sense. But if the second repo fails to clone, it might be because:
- The **wrong key gets offered first**, and GitHub rejects the rest (a common SSH behavior),
- Or there’s something about how Dagger forwards the agent that doesn’t handle multiple keys well — especially if `IdentitiesOnly` isn’t enforced.

### 🧩 To help narrow this down, a couple of things would be useful to know:
1. Have you tried cloning both repos manually (outside Dagger) with the same setup to confirm the keys themselves work? (it seems so, just wanna triple check 🙏)
3. Are you using an `~/.ssh/config` file with something like `IdentitiesOnly yes` and `IdentityFile` set per host alias?   - That can help pin each key to the right repo and avoid ambiguity. Even better: could you please share that `.ssh/config` file (anonymized) ? I was never able to reproduce on my end -- and it's probably an ssh config thing-y that we don't properly forward nor mount 👀 

Let us know what you find — this kind of multi-key setup is tricky, but we definitely want to get it working smoothly 🙏

**Comment by ielsharef** [comment]
Hey @grouville, thank you for your quick response🙏

1. I've cloned both repos manually outside of Dagger with the same setup to confirm the keys themselves work
2. I've tried an ~/.ssh/config file with the contents below, but no luck. I might be overlooking something - Dagger seems to have difficulty resolving / loading my module when I do this

```
Host github-daggerverse
  HostName github.com
  User git
  IdentityFile ~/.ssh/daggerverse_deploy_key
  IdentitiesOnly yes
Host github-templates
  HostName github.com
  User git
  IdentityFile ~/.ssh/templates_deploy_key
  IdentitiesOnly yes
```

Example call
```
dagger -m git@github-daggerverse:ielsharef/8288-daggerverse.git/example-module@main call test-arg-clone \
    --repo-arg=git@github-templates:ielsharef/8288-templates.git
```
  
To make sure this wasn't just local setup weirdness on my end, I created a private test environment within GitHub. I've given you access to these repos if you're comfortable looking there directly

- [daggerverse repo ](https://github.com/ielsharef/8288-daggerverse) has a deploy key configured for access and an example dagger module that takes a *dagger.Directory arg
- [templates repo](https://github.com/ielsharef/8288-templates) has a deploy key configured for access and is the *dagger.Directory arg passed to the dagger module
- [workflows repo](https://github.com/ielsharef/8288-workflows) relies on deploy keys from the daggerverse repo and the template repo to make the dagger module call. Each deploy key is read-only and only gives access to its specific repo

Example runs: 
- Without host aliases, the module can be called but fails on cloning the *dagger.Directory arg 
https://github.com/ielsharef/8288-workflows/actions/runs/14541756960/job/40800990089#step:3:738

- With host aliases, it's not able to load the module
https://github.com/ielsharef/8288-workflows/actions/runs/14541664566/job/40800718038#step:3:569

Sorry for the delay, this got buried on my end 🙂

**Comment by ielsharef** [comment]
@grouville is there anything more I can share to help with this? I know replicating it can be tricky and it may be something I'm misunderstanding in setup

Thanks again for looking into this 🙏

**Comment by ielsharef** [comment]
@grouville @aweris bumping in case this got buried 🙂

As an update, I've also tried testing with v0.18.6 using the *dagger.Directory argument for functions vs using dag.Git with the ssh socket explicitly passed in, but still no luck. Is there anything else you'd recommend I try in case I'm overlooking something?

**Comment by grouville** [comment]
> [@grouville](https://github.com/grouville) [@aweris](https://github.com/aweris) bumping in case this got buried 🙂
> 
> As an update, I've also tried testing with v0.18.6 using the *dagger.Directory argument for functions vs using dag.Git with the ssh socket explicitly passed in, but still no luck. Is there anything else you'd recommend I try in case I'm overlooking something?

Hey sorry @ielsharef I got sidetracked -- I promise to take a look after wednesday ahah 🙏

**Comment by grouville** [comment]
Hey @ielsharef I'm actually looking at it 👀 Could you please re-invite me to the repos ? I forgot it had an expiration time

**Comment by ielsharef** [comment]
@grouville no worries! I've just re-invited you, and you'll have write access by default to test things. 

The run below uses dagger v0.18.6
https://github.com/ielsharef/8288-workflows/actions/runs/15098287985/job/42435882484#step:3:716

It includes blocks to call both without host aliases (80-82) and with aliases (84-86). I was thinking it might help for easier testing to be able to have both defined and just comment out the first block when trying to test with aliases
http://github.com/ielsharef/8288-workflows/blob/main/.github/workflows/templates.yml#L80

Please let me know if you have any questions, and thanks again for looking into this 🙏
